File: geo/citability.py
"""Page-level citability — how quotable RISA's own pages are to an answer engine.

Ports the geo-citability rubric into a deterministic, network-only scorer (no
LLM, no cost): fetch the brand's homepage, follow a few same-domain links, and
score each page 0-100 on five sub-signals an AI weighs when deciding to quote a
passage:

    answer-first      0.30   a direct, declarative lead (not a marketing tease)
    self-containment  0.25   passages that stand alone (low pronoun-opener rate)
    structure         0.20   headings + lists the model can lift cleanly
    statistics        0.15   hard numbers (%, $, figures) per 1k words
    specificity       0.10   named entities / proper nouns (stand-in for unique)

Writes output/citability.json. Pure stdlib (urllib + re).
"""
from __future__ import annotations

import json
import os
import re
import urllib.request
from datetime import datetime
from html.parser import HTMLParser
from urllib.parse import urljoin, urlparse

from .config import AppConfig

_UA = "Mozilla/5.0 (compatible; RISA-GEO-Citability/1.0)"
_MAX_PAGES = 6


def _fetch(url: str, timeout: int = 12):
    try:
        req = urllib.request.Request(url, headers={"User-Agent": _UA})
        with urllib.request.urlopen(req, timeout=timeout) as r:
            ctype = r.headers.get("Content-Type", "")
            if "html" not in ctype and ctype:
                return None, None
            return r.status, r.read().decode("utf-8", "replace")
    except Exception:
        return None, None


class _Extractor(HTMLParser):
    """Pull visible text, headings, list items, and same-domain links."""

    SKIP = {"script", "style", "noscript", "svg", "head", "nav", "footer"}

    def __init__(self):
        super().__init__()
        self.text: list[str] = []
        self.h2 = 0
        self.h3 = 0
        self.li = 0
        self.links: list[str] = []
        self.title = ""
        self._skip = 0
        self._tag = ""

    def handle_starttag(self, tag, attrs):
        self._tag = tag
        if tag in self.SKIP:
            self._skip += 1
        if tag == "h2":
            self.h2 += 1
        elif tag == "h3":
            self.h3 += 1
        elif tag == "li":
            self.li += 1
        elif tag == "a":
            for k, v in attrs:
                if k == "href" and v:
                    self.links.append(v)

    def handle_endtag(self, tag):
        if tag in self.SKIP and self._skip:
            self._skip -= 1

    def handle_data(self, data):
        if self._skip:
            return
        s = data.strip()
        if not s:
            return
        if self._tag == "title" and not self.title:
            self.title = s
        self.text.append(s)


_PRONOUN_OPENERS = ("it ", "this ", "that ", "they ", "these ", "those ", "he ", "she ", "we ", "our ", "its ")
_STAT = re.compile(r"(\$\s?\d[\d,.]*|\d[\d,.]*\s?%|\b\d{2,}\b|\b\d+\.\d+\b)")
_PROPER = re.compile(r"\b[A-Z][a-zA-Z]+(?:\s[A-Z][a-zA-Z]+)*\b")


def _score_page(url: str, html: str) -> dict | None:
    p = _Extractor()
    try:
        p.feed(html)
    except Exception:
        return None
    blob = " ".join(p.text)
    words = blob.split()
    n_words = len(words)
    if n_words < 120:
        return None  # thin / non-content page — don't score it

    sentences = [s.strip() for s in re.split(r"(?<=[.!?])\s+", blob) if len(s.strip()) > 24]
    lead = " ".join(words[:60]).lower()

    # answer-first: lead is declarative + on-topic, not a pronoun/CTA tease
    answer = 100.0
    if lead.startswith(_PRONOUN_OPENERS):
        answer -= 35
    if not _STAT.search(" ".join(words[:90])) and "is " not in lead and "are " not in lead:
        answer -= 25
    if any(w in lead for w in ("welcome", "sign up", "get started", "book a demo", "contact us")):
        answer -= 25
    answer = max(0.0, answer)

    # self-containment: share of sentences that DON'T open with a bare pronoun
    if sentences:
        bad = sum(1 for s in sentences if s.lower().startswith(_PRONOUN_OPENERS))
        self_contained = 100 * (1 - bad / len(sentences))
    else:
        self_contained = 40.0

    # structure: headings + lists relative to length
    per_k = 1000 / max(n_words, 1)
    headings = (p.h2 + p.h3) * per_k
    lists = p.li * per_k
    structure = min(100.0, 55 * min(1.0, headings / 6) + 45 * min(1.0, lists / 10))

    # statistics density
    stats_n = len(_STAT.findall(blob))
    statistics = min(100.0, 100 * min(1.0, (stats_n * per_k) / 8))

    # specificity: distinct proper-noun phrases (entity richness)
    proper = len(set(_PROPER.findall(blob)))
    specificity = min(100.0, 100 * min(1.0, (proper * per_k) / 12))

    score = round(
        0.30 * answer + 0.25 * self_contained + 0.20 * structure
        + 0.15 * statistics + 0.10 * specificity
    )
    return {
        "url": url,
        "title": (p.title or url)[:90],
        "words": n_words,
        "score": score,
        "subscores": {
            "answer_first": round(answer),
            "self_contained": round(self_contained),
            "structure": round(structure),
            "statistics": round(statistics),
            "specificity": round(specificity),
        },
        "stats_found": stats_n,
        "headings": p.h2 + p.h3,
    }


def _discover(base: str, html: str) -> list[str]:
    """Homepage + a few same-domain content links."""
    host = urlparse(base).netloc
    p = _Extractor()
    try:
        p.feed(html)
    except Exception:
        pass
    seen, urls = set(), [base]
    skip_frag = ("#", "mailto:", "tel:", ".pdf", ".png", ".jpg", ".svg", "/privacy", "/terms", "/cookie")
    for href in p.links:
        u = urljoin(base, href.split("#")[0])
        if urlparse(u).netloc != host or u in seen or u == base:
            continue
        if any(s in u.lower() for s in skip_frag):
            continue
        seen.add(u)
        urls.append(u)
        if len(urls) >= _MAX_PAGES:
            break
    return urls


def citability(cfg: AppConfig) -> dict:
    base = "https://" + cfg.brand.domain.rstrip("/")
    print(f"[citability] scoring owned pages under {base}…")
    status, html = _fetch(base)
    pages: list[dict] = []
    if status == 200 and html:
        for url in _discover(base, html):
            st, body = _fetch(url) if url != base else (status, html)
            if st == 200 and body:
                sc = _score_page(url, body)
                if sc:
                    pages.append(sc)
    pages.sort(key=lambda x: x["score"], reverse=True)
    score = round(sum(p["score"] for p in pages) / len(pages)) if pages else 0

    out = {
        "domain": base,
        "fetched_at": datetime.now().isoformat(timespec="seconds"),
        "score": score,
        "pages_scored": len(pages),
        "rubric": {
            "answer_first": 0.30, "self_contained": 0.25, "structure": 0.20,
            "statistics": 0.15, "specificity": 0.10,
        },
        "pages": pages,
        "note": "deterministic (no-LLM) page citability — uniqueness uses entity richness as a stand-in",
    }
    os.makedirs(cfg.output_dir, exist_ok=True)
    with open(os.path.join(cfg.output_dir, "citability.json"), "w", encoding="utf-8") as f:
        json.dump(out, f, ensure_ascii=False, indent=2)
    print(f"[citability] mean score {score}/100 across {len(pages)} pages")
    return out

File: web/lib/pipeline/citability.ts
// Page-level citability — how quotable the brand's own pages are to an answer engine.
// Ported from geo/citability.py (deterministic, no-LLM rubric).
import path from "node:path";
import { AppConfig } from "./config";
import { fetchText, extractHtml, resolveUrl, type Extracted } from "./html";
import { writeJson } from "./io";
import type { Progress } from "./collect";

const MAX_PAGES = 6;
const PRONOUN_OPENERS = ["it ", "this ", "that ", "they ", "these ", "those ", "he ", "she ", "we ", "our ", "its "];
const STAT = /(\$\s?\d[\d,.]*|\d[\d,.]*\s?%|\b\d{2,}\b|\b\d+\.\d+\b)/g;
const PROPER = /\b[A-Z][a-zA-Z]+(?:\s[A-Z][a-zA-Z]+)*\b/g;

function nowIso(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "");
}

function startsWithPronoun(s: string): boolean {
  const low = s.toLowerCase();
  return PRONOUN_OPENERS.some((p) => low.startsWith(p));
}

export function scorePage(url: string, html: string): any | null {
  let p: Extracted;
  try {
    p = extractHtml(html);
  } catch {
    return null;
  }
  const blob = p.text.join(" ");
  const words = blob.split(/\s+/).filter(Boolean);
  const nWords = words.length;
  if (nWords < 120) return null;

  const sentences = blob
    .split(/(?<=[.!?])\s+/)
    .map((s) => s.trim())
    .filter((s) => s.length > 24);
  const lead = words.slice(0, 60).join(" ").toLowerCase();

  // answer-first
  let answer = 100.0;
  if (startsWithPronoun(lead)) answer -= 35;
  const first90 = words.slice(0, 90).join(" ");
  if (!STAT.test(first90) && !lead.includes("is ") && !lead.includes("are ")) answer -= 25;
  STAT.lastIndex = 0;
  if (["welcome", "sign up", "get started", "book a demo", "contact us"].some((w) => lead.includes(w))) answer -= 25;
  answer = Math.max(0, answer);

  // self-containment
  let selfContained = 40.0;
  if (sentences.length) {
    const bad = sentences.filter((s) => startsWithPronoun(s)).length;
    selfContained = 100 * (1 - bad / sentences.length);
  }

  // structure
  const perK = 1000 / Math.max(nWords, 1);
  const headings = (p.h2 + p.h3) * perK;
  const lists = p.li * perK;
  const structure = Math.min(100, 55 * Math.min(1, headings / 6) + 45 * Math.min(1, lists / 10));

  // statistics density
  const statsN = (blob.match(STAT) || []).length;
  STAT.lastIndex = 0;
  const statistics = Math.min(100, 100 * Math.min(1, (statsN * perK) / 8));

  // specificity
  const proper = new Set(blob.match(PROPER) || []).size;
  const specificity = Math.min(100, 100 * Math.min(1, (proper * perK) / 12));

  const score = Math.round(
    0.3 * answer + 0.25 * selfContained + 0.2 * structure + 0.15 * statistics + 0.1 * specificity,
  );
  return {
    url,
    title: (p.title || url).slice(0, 90),
    words: nWords,
    score,
    subscores: {
      answer_first: Math.round(answer),
      self_contained: Math.round(selfContained),
      structure: Math.round(structure),
      statistics: Math.round(statistics),
      specificity: Math.round(specificity),
    },
    stats_found: statsN,
    headings: p.h2 + p.h3,
  };
}

export function discover(base: string, html: string): string[] {
  let host: string;
  try {
    host = new URL(base).host;
  } catch {
    host = "";
  }
  let p: Extracted;
  try {
    p = extractHtml(html);
  } catch {
    p = { text: [], h2: 0, h3: 0, li: 0, links: [], title: "" };
  }
  const seen = new Set<string>();
  const urls = [base];
  const skipFrag = ["#", "mailto:", "tel:", ".pdf", ".png", ".jpg", ".svg", "/privacy", "/terms", "/cookie"];
  for (const href of p.links) {
    const u = resolveUrl(base, href.split("#")[0]);
    let uHost: string;
    try {
      uHost = new URL(u).host;
    } catch {
      continue;
    }
    if (uHost !== host || seen.has(u) || u === base) continue;
    if (skipFrag.some((s) => u.toLowerCase().includes(s))) continue;
    seen.add(u);
    urls.push(u);
    if (urls.length >= MAX_PAGES) break;
  }
  return urls;
}

export async function citability(cfg: AppConfig, onProgress?: Progress): Promise<any> {
  const base = "https://" + cfg.brand.domain.replace(/\/$/, "");
  onProgress?.(`[citability] scoring owned pages under ${base}…`);
  const { status, body } = await fetchText(base, { htmlOnly: true });
  const pages: any[] = [];
  if (status === 200 && body) {
    for (const url of discover(base, body)) {
      const { status: st, body: pageBody } = url === base ? { status, body } : await fetchText(url, { htmlOnly: true });
      if (st === 200 && pageBody) {
        const sc = scorePage(url, pageBody);
        if (sc) pages.push(sc);
      }
    }
  }
  pages.sort((a, b) => b.score - a.score);
  const score = pages.length ? Math.round(pages.reduce((a, p) => a + p.score, 0) / pages.length) : 0;

  const out = {
    domain: base,
    fetched_at: nowIso(),
    score,
    pages_scored: pages.length,
    rubric: { answer_first: 0.3, self_contained: 0.25, structure: 0.2, statistics: 0.15, specificity: 0.1 },
    pages,
    note: "deterministic (no-LLM) page citability — uniqueness uses entity richness as a stand-in",
  };
  writeJson(path.join(cfg.outputDir, "citability.json"), out);
  onProgress?.(`[citability] mean score ${score}/100 across ${pages.length} pages`);
  return out;
}
